syntax = "proto3";

package envoy.service.auth.v3;

import "google/protobuf/struct.proto";

// Envoy ext_authz API를 위한 외부 인증 서비스 정의입니다.
service Authorization {
  // Envoy가 요청 승인 여부를 확인하기 위해 호출하는 RPC입니다.
  rpc Check(CheckRequest) returns (CheckResponse);
}

// 클라이언트(Envoy)에서 전달하는 체크 요청 메시지입니다.
message CheckRequest {
  // 임의의 속성을 담을 수 있는 필드입니다.
  google.protobuf.Struct attributes = 1;
}

// 체크 응답 메시지입니다.
message CheckResponse {
  // 응답 결과를 oneof로 구분합니다.
  oneof result {
    OkResponse ok_response = 1;
    DeniedResponse denied_response = 2;
  }
}

// 요청을 승인하는 응답 메시지입니다.
message OkResponse {
  // 응답 헤더(추가 헤더 삽입 등)를 담을 수 있습니다.
  google.protobuf.Struct headers = 1;
}

// 요청을 거부하는 응답 메시지입니다.
message DeniedResponse {
  // HTTP 상태 코드 (예: 403 등)
  int32 status = 1;
  // 응답 헤더
  google.protobuf.Struct headers = 2;
  // 거부 사유 메시지
  string body = 3;
}

